File: go-backend-interview-prep/01_go_basics/10_named_returns_and_variadics/main.py
# Python equivalent of Named Returns and Variadics (compare with main.go)
# Python has no named returns. You just return values.
# Python has *args for variadics.


def split(total: int) -> tuple[int, int]:
    """Go uses named returns; Python just returns a tuple."""
    half = total // 2
    remainder = total % 2
    return half, remainder


def parse_coords(input_str: str) -> tuple[float, float, str | None]:
    """Named returns in Go serve as docs. In Python, use type hints."""
    lat = 35.6762
    lng = 139.6503
    return lat, lng, None


def sum_all(*nums: int) -> int:
    """Go: func sum(nums ...int). Python: *args."""
    total = 0
    for n in nums:
        total += n
    return total


def greet_all(greeting: str, *names: str) -> None:
    """Go: func greetAll(greeting string, names ...string)."""
    for name in names:
        print(greeting, name)


def main():
    # --- "Named" returns (just tuple unpacking) ---
    h, r = split(17)
    print(f"split(17): half={h} remainder={r}")

    lat, lng, _ = parse_coords("ignored")
    print(f"coords: lat={lat:.4f} lng={lng:.4f}")

    # --- Variadic: individual args ---
    print("sum(1,2,3):", sum_all(1, 2, 3))
    print("sum(10,20):", sum_all(10, 20))
    print("sum():", sum_all())  # valid -- returns 0

    # --- Variadic: spread a list ---
    scores = [90, 85, 92, 78]
    print("sum(*scores):", sum_all(*scores))  # Go: sum(scores...)

    # --- Variadic with fixed param ---
    greet_all("Hello", "Alice", "Bob", "Charlie")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/01_go_basics/10_named_returns_and_variadics/main.go
package main

import "fmt"

// --- Named return values ---
func split(total int) (half, remainder int) {
	half = total / 2
	remainder = total % 2
	return // naked return: returns half and remainder
}

// --- Named returns for documentation ---
func parseCoords(input string) (lat, lng float64, err error) {
	// In real code you would parse the string.
	// Named returns make the API self-documenting.
	lat = 35.6762
	lng = 139.6503
	return
}

// --- Variadic function ---
func sum(nums ...int) int {
	total := 0
	for _, n := range nums {
		total += n
	}
	return total
}

// --- Variadic with a fixed first param ---
func greetAll(greeting string, names ...string) {
	for _, name := range names {
		fmt.Println(greeting, name)
	}
}

func main() {
	// --- Named returns ---
	h, r := split(17)
	fmt.Printf("split(17): half=%d remainder=%d\n", h, r)

	lat, lng, _ := parseCoords("ignored")
	fmt.Printf("coords: lat=%.4f lng=%.4f\n", lat, lng)

	// --- Variadic: individual args ---
	fmt.Println("sum(1,2,3):", sum(1, 2, 3))
	fmt.Println("sum(10,20):", sum(10, 20))
	fmt.Println("sum():", sum()) // valid -- returns 0

	// --- Variadic: spread a slice ---
	scores := []int{90, 85, 92, 78}
	fmt.Println("sum(scores...):", sum(scores...))

	// --- Variadic with fixed param ---
	greetAll("Hello", "Alice", "Bob", "Charlie")
}
